File: packet_factory.py
import json
from protocol import PacketType

class PacketFactory:
    @staticmethod
    def build(packet_id: int, data: dict) -> str:
        """Serialize a packet to a JSON string."""
        packet = {"id": packet_id, "data": data}
        return json.dumps(packet)

    @staticmethod
    def parse(raw_data: str):
        """Deserialize raw packet data into a dict."""
        try:
            packet = json.loads(raw_data)
            return packet.get("id"), packet.get("data")
        except json.JSONDecodeError:
            return None, None

File: client/PacketFactory.cs
using UnityEngine;

[System.Serializable]
public class Packet<T>
{
    public int id;
    public T data;
}

public static class PacketFactory
{
    public static string Build<T>(int id, T data)
    {
        Packet<T> packet = new Packet<T> { id = id, data = data };
        return JsonUtility.ToJson(packet) + "\n"; // newline for TCP delimiting
    }

    public static Packet<T> Parse<T>(string json)
    {
        return JsonUtility.FromJson<Packet<T>>(json);
    }
}
